### DISCUSSION TITLE
General limitation for mounting WEKA volume

### DISCUSSION THREAD

**Discussion by jslouisyou** [weka/csi-wekafs#193] [Q&A] [answered]
Hello,

Are there any general limitations for mounting WEKA volume in k8s? I can't find any of these in kubernetes documentations or anywhere. 

For example,
1. The maximum number of PersistentVolume or PersistentVolumeClaim can be created within same StorageClass across all nodes, is limited?
2. The maximum number of mounted filesystem limits within one node limited? (I'm using Ubuntu 20.04)

Thanks!

**Comment by sergeyberezansky** [comment, marked as answer]
Hello @jslouisyou ,

1. Maxumum number of PVCs depends on backing type and version of Weka cluster software. Please refer to this document for additional information: https://github.com/weka/csi-wekafs/blob/main/docs/usage.md
2. Maximum number of mounted filesystems is not set by the CSI plugin. For directory-backed and snapshot-backed volumes, a single filesystem mount is made by the node. For filesystem-backed volumes, obviously, a single mount is done per each filesystem.
So basically, even using a single filesystem mount, a single node could publish thousands of different PersistentVolumes.
Please refer to Weka documentation for additional limitations that could apply.

**Reply by jslouisyou** [reply]
Thanks @sergeyberezansky .
Following WEKA documentation as you mentioned, I'm now using WEKA backend 4.0 with directory-backend volume type.
FYI: I'm using k8s v1.21.6 and csi-wekafs plugin v2.3.0

Refer to usage, I can create unlimited number of volumes (PV, PVC I think) and it seems obvious.

But recently I faced an issue that there are some PODs that cannot mount WEKA volume even if filesystem is set to `ReadWriteMany` and there were 15,000 PV and PVCs which referred just one volume. (I'm sorry but there aren't any logs left from csi-wekafs-node PODs)

And I'm wondering:
1. In terms of `ReadWriteMany`, is there any limitation that forbid the number of PODs that share same volume in csi-wekafs plugin level? If not, then unlimited number of PODs can share just one volume?
2. In terms of csi-wekafs, is there any limitation or policy that set the maximum number of mounts from one filesystem (`StorageClass`)?

Thanks.

**Reply by sergeyberezansky** [reply]
There is no limitation whatsoever on the Weka CSI plugin binaries.
I am not sure about limitations applied by Kubernetes or CSI node publisher sidecar.
Need to check this.

It would be really nice if you could provide logs of the relevant csi-wekafs-node pod from all its containers.
Please open a new issue if it persists and agtach the logs.
Thank you.

**Reply by sergeyberezansky** [reply]
Wekafs logs can be easily filtered by trace_id for each NodePublishVolumeRequest.
Sidecar logs are not using opentelemetry, so it might be more complicated in scale.

But please try to gather the logs and we will look together.

**Reply by sergeyberezansky** [reply]
@jslouisyou if you believe the question was answered, please mark it. 
Feel free to reach out if needed!

**Comment by jslouisyou** [comment]
@sergeyberezansky Thanks for your help!
